### PR TITLE
Create single table synthesis property

### LINKED ISSUE #390: Create single table synthesis property
### Problem Description

Now that we have property classes, we want to move the metrics used in the `DiagnosticReport` into their respective properties.

This is the spec for the `Synthesis` property

#### Attributes

- `_details` - A dataframe containing the following columns
    - 'Metric'
    - 'Score'
    - 'Num Matched Rows'
    - 'Num New Rows'

The metrics used in this class are: [NewRowSynthesis]

#### Methods

- `_generate_details(self, real_data, synthetic_data, metadata, progress_bar)` - Generate the `_details` dataframe that will be used by the base class to get the score.
Pseudocode:
```python
score = NewRowSynthesis.compute_breakdown(real_data, synthetic_data, synthetic_sample_size: 10000)
```
- `get_visualization()` - Returns a plotly.graph_objects._figure.Figure object for the specified table. Use code similar to what's in [get_synthesis_plot](https://github.com/sdv-dev/SDMetrics/blob/master/sdmetrics/reports/single_table/plot_utils.py#L390).

### Expected behavior

```python
>>> prop = Synthesis()
>>> prop.get_score()
>>> 0.89
>>> prop.get_visualization()
```

### Additional context

- we should catch any exceptions the same way we do [here](https://github.com/sdv-dev/SDMetrics/blob/master/sdmetrics/reports/single_table/_properties/column_shapes.py)

### PR DESCRIPTION
**Pull request by R-Palazzo**
Resolve #390

### CODE REVIEW

File: sdmetrics/reports/single_table/_properties/synthesis.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    """
+
+    metric = NewRowSynthesis
+
+    def _generate_details(self, real_data, synthetic_data, metadata, progress_bar=None):
+        """Generate the _details dataframe for the synthesis property.
+
+        Args:
+            real_data (pandas.DataFrame):
+                The real data
```
Comment (fealho) [inline sdmetrics/reports/single_table/_properties/synthesis.py:24]: We usually end args with dots. Same for args below.

File: sdmetrics/reports/single_table/_properties/synthesis.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                The metadata of the table
+            progress_bar (tqdm.tqdm or None):
+                The progress bar to use. Defaults to tqdm.
+
+        Returns:
+            pandas.DataFrame
+        """
+        name = self.metric.__name__
+        error_message = np.nan
+        if len(synthetic_data) > 10000:
```
Comment (fealho) [inline sdmetrics/reports/single_table/_properties/synthesis.py:37]: Why is this here? And should this not log some sort of warning?

It can also be written as one line: `sample_size = len(synthetic_data) if len(synthetic_data) <= 10000 else 10000`

File: sdmetrics/reports/single_table/_properties/synthesis.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            score = np.nan
+            num_matched_rows = np.nan
+            num_new_rows = np.nan
+            error_message = f'Error: {type(e).__name__} {e}'
+
+        finally:
+            if progress_bar:
+                progress_bar.update()
+
+        result = pd.DataFrame({
```
Comment (fealho) [inline sdmetrics/reports/single_table/_properties/synthesis.py:60]: Are any of the objects a list? If not, it's weird to me we are using a DataFrame when a simple dict would suffice.

File: sdmetrics/reports/single_table/_properties/synthesis.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            'Metric': name,
+            'Score': score,
+            'Num Matched Rows': num_matched_rows,
+            'Num New Rows': num_new_rows,
+            'Error': error_message,
+        }, index=[0])
+
+        if result['Error'].isna().all():
```
Comment (fealho) [inline sdmetrics/reports/single_table/_properties/synthesis.py:68]: Can result['Error'] be a list? If not we don't need the `all`.

I also find it more intuitive to add the errors to `result` if there are any, rather than removing them if there were none.

File: sdmetrics/reports/single_table/_properties/synthesis.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        return result
+
+    def get_visualization(self):
+        """Create a plot to show the synthesis property.
+
+        Returns:
+            plotly.graph_objects._figure.Figure
+        """
+        labels = ['Exact Matches', 'Novel Rows']
+        values = list(self._details[['Num Matched Rows', 'Num New Rows']].to_numpy()[0])
```
Comment (fealho) [inline sdmetrics/reports/single_table/_properties/synthesis.py:80]: I don't think you need `to_numpy`, you can select the first row with `iloc[0]`.

File: sdmetrics/reports/single_table/_properties/synthesis.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+        average_score = self._compute_average()
+        if not np.isnan(average_score):
```
Comment (fealho) [inline sdmetrics/reports/single_table/_properties/synthesis.py:83]: I don't think you need this check, `round` works for nans.

File: tests/unit/reports/single_table/_properties/test_synthesis.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                'col4': {'sdtype': 'datetime'}
+            }
+        }
+
+        newrowsynthesis_mock.return_value = {
+            'score': 0.25,
+            'num_matched_rows': 3,
+            'num_new_rows': 1,
+        }
+        # Run
```
Comment (fealho) [inline tests/unit/reports/single_table/_properties/test_synthesis.py:41]: New line

File: tests/unit/reports/single_table/_properties/test_synthesis.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class TestSynthesis:
+
+    @patch('sdmetrics.reports.single_table._properties.synthesis.'
+           'NewRowSynthesis.compute_breakdown')
+    def test__generate_details(self, newrowsynthesis_mock):
+        """Test the ``_generate_details`` method."""
+        # Setup
+        real_data = pd.DataFrame({
```
Comment (fealho) [inline tests/unit/reports/single_table/_properties/test_synthesis.py:15]: I don't think you need the real_data/metadata block, since they are not used for anything (just pass empty dataframes or mocks), and synthetic_data can just be a one liner, since it only checks the size of that. Just to make it more obvious that what's actually been tested is the return of the mock.

File: tests/unit/reports/single_table/_properties/test_synthesis.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        # Run
+        synthesis_property = Synthesis()
+        details = synthesis_property._generate_details(real_data, synthetic_data, metadata)
+
+        # Assert
+        expected_calls = [
+            call(real_data, synthetic_data, synthetic_sample_size=4)
+        ]
+
+        newrowsynthesis_mock.assert_has_calls(expected_calls)
```
Comment (fealho) [inline tests/unit/reports/single_table/_properties/test_synthesis.py:50]: You also need to test when synthetic data has over 10,000 rows.

**Review by fealho — CHANGES_REQUESTED** [review]
Looking good, just a few suggestions.

File: sdmetrics/reports/single_table/_properties/synthesis.py (hunk 1 above)
Comment (R-Palazzo) [reply sdmetrics/reports/single_table/_properties/synthesis.py]: Yes, done in [b00d7ca](https://github.com/sdv-dev/SDMetrics/pull/398/commits/b00d7cafd54bf24f3b76c78763bddbd65d5b9408)

File: sdmetrics/reports/single_table/_properties/synthesis.py (hunk 2 above)
Comment (R-Palazzo) [reply sdmetrics/reports/single_table/_properties/synthesis.py:37]: @npatki I was also not sure about this.
For the `NewRowSynthesis`, by default, we limit the `sample size` of the synthetic data to 10000 right?
Should we raise a log or a warning if the synthetic data is larger than 10000 rows?

File: sdmetrics/reports/single_table/_properties/synthesis.py (hunk 3 above)
Comment (R-Palazzo) [reply sdmetrics/reports/single_table/_properties/synthesis.py:60]: It's to keep the same abstraction as the other properties. `_generate_details` should generate the details table, which is a DataFrame. I agree it's a bit weird because there is only 1 row.

File: sdmetrics/reports/single_table/_properties/synthesis.py (hunk 5 above)
Comment (R-Palazzo) [reply sdmetrics/reports/single_table/_properties/synthesis.py]: Yes, done in [b00d7ca](https://github.com/sdv-dev/SDMetrics/pull/398/commits/b00d7cafd54bf24f3b76c78763bddbd65d5b9408)

File: sdmetrics/reports/single_table/_properties/synthesis.py (hunk 6 above)
Comment (R-Palazzo) [reply sdmetrics/reports/single_table/_properties/synthesis.py]: Yes, done in [b00d7ca](https://github.com/sdv-dev/SDMetrics/pull/398/commits/b00d7cafd54bf24f3b76c78763bddbd65d5b9408)

File: tests/unit/reports/single_table/_properties/test_synthesis.py (hunk 7 above)
Comment (R-Palazzo) [reply tests/unit/reports/single_table/_properties/test_synthesis.py:31]: Yes, done in [b00d7ca](https://github.com/sdv-dev/SDMetrics/pull/398/commits/b00d7cafd54bf24f3b76c78763bddbd65d5b9408)

File: tests/unit/reports/single_table/_properties/test_synthesis.py (hunk 8 above)
Comment (R-Palazzo) [reply tests/unit/reports/single_table/_properties/test_synthesis.py]: Yes, done in [b00d7ca](https://github.com/sdv-dev/SDMetrics/pull/398/commits/b00d7cafd54bf24f3b76c78763bddbd65d5b9408)

File: tests/unit/reports/single_table/_properties/test_synthesis.py (hunk 9 above)
Comment (R-Palazzo) [reply tests/unit/reports/single_table/_properties/test_synthesis.py:42]: Yes, done in [b00d7ca](https://github.com/sdv-dev/SDMetrics/pull/398/commits/b00d7cafd54bf24f3b76c78763bddbd65d5b9408)

File: sdmetrics/reports/single_table/_properties/synthesis.py (hunk 2 above)
Comment (amontanez24) [reply sdmetrics/reports/single_table/_properties/synthesis.py:37]: I don't think we need this here. Just pass it in as 10000 to the metric. If the data has more than 10000 rows it will subsample, otherwise it will use all the data

File: tests/unit/reports/single_table/_properties/test_synthesis.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import pandas as pd
+
+from sdmetrics.reports.single_table._properties.synthesis import Synthesis
+
+
+class TestSynthesis:
+
+    @patch('sdmetrics.reports.single_table._properties.synthesis.'
+           'NewRowSynthesis.compute_breakdown')
+    def test__generate_details(self, newrowsynthesis_mock):
```
Comment (amontanez24) [inline tests/unit/reports/single_table/_properties/test_synthesis.py:12]: can we add a unit test for the error case as well

**Review by amontanez24 — CHANGES_REQUESTED** [review]

File: tests/unit/reports/single_table/_properties/test_synthesis.py (hunk 10 above)
Comment (R-Palazzo) [reply tests/unit/reports/single_table/_properties/test_synthesis.py:13]: Yes I was doing it hahaha. Done in [a310a75](https://github.com/sdv-dev/SDMetrics/pull/398/commits/a310a7541be76a3cc5fac3b80250bbc1b7f8c665)

File: sdmetrics/reports/single_table/_properties/synthesis.py (hunk 2 above)
Comment (R-Palazzo) [reply sdmetrics/reports/single_table/_properties/synthesis.py:37]: I did this not to raise this warning (because the user no longer has the possibility to adjust the `sample_size`): https://github.com/sdv-dev/SDMetrics/blob/ddec4f76e3cb866d13b8aded0e319b3e4e96ca3d/sdmetrics/single_table/new_row_synthesis.py#L68

Comment (amontanez24) [reply sdmetrics/reports/single_table/_properties/synthesis.py:37]: Or you can leave it but make it onle line like @fealho suggested

Comment (R-Palazzo) [reply sdmetrics/reports/single_table/_properties/synthesis.py]: Yes alright, done in [4669397](https://github.com/sdv-dev/SDMetrics/pull/398/commits/4669397521b9df075bc08a7d8a4fbbd880177250)

**Review by amontanez24 — APPROVED** [review]
LGTM!

File: tests/unit/reports/single_table/_properties/test_synthesis.py [hunk 11]
```diff
@@ -50,6 +51,38 @@ def test__generate_details(self, newrowsynthesis_mock):
 
         pd.testing.assert_frame_equal(details, expected__details)
 
+    @patch('sdmetrics.reports.single_table._properties.synthesis.'
+           'NewRowSynthesis.compute_breakdown')
+    def test__generate_details_error(self, newrowsynthesis_mock):
+        """Test the ``_generate_details`` method when the metric raises an error."""
+        # Setup
+
```
Comment (fealho) [inline tests/unit/reports/single_table/_properties/test_synthesis.py:59]: Delete new line

**Review by fealho — APPROVED** [review]
Thanks for addressing all those comments!
